File: solutions/haskell/list-ops/1/src/ListOps.hs
{-# LANGUAGE BangPatterns #-}

module ListOps (
  length,
  reverse,
  map,
  filter,
  foldr,
  foldl',
  (++),
  concat,
) where

import Prelude hiding (
  concat,
  filter,
  foldr,
  length,
  map,
  reverse,
  (++),
 )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ !z [] = z
foldl' f z (x : xs) = foldl' f (f z x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f z xs = foldl' (flip f) z (reverse xs)

length :: [a] -> Int
length = go 0
 where
  go acc [] = acc
  go acc (_ : xs) = go (acc + 1) xs

reverse :: [a] -> [a]
reverse = go []
 where
  go acc [] = acc
  go acc (x : xs) = go (x : acc) xs

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x : xs) = f x : map f xs

filter :: (a -> Bool) -> [a] -> [a]
filter _ [] = []
filter p (x : xs)
  | p x = x : filter p xs
  | otherwise = filter p xs

(++) :: [a] -> [a] -> [a]
[] ++ ys = ys
(x : xs) ++ ys = x : (xs ++ ys)

concat :: [[a]] -> [a]
concat = foldr (++) []

File: solutions/lua/list-ops/1/list-ops.lua
local function append(xs, ys)
  for _, v in ipairs(ys) do table.insert(xs, v) end
  return xs
end

local function concat(...)
  local result = {}
  for _, v in ipairs(table.pack(...)) do result = append(result, v) end
  return result
end

local function reverse(xs)
  local result = {}
  for _, v in ipairs(xs) do table.insert(result, 1, v) end
  return result
end

local function foldl(xs, value, f)
  if #xs == 0 then return value end
  return foldl({ table.unpack(xs, 2) }, f(value, xs[1]), f)
end

local function map(xs, f)
  local result = {}
  for _, v in ipairs(xs) do table.insert(result, f(v)) end
  return result
end

local function filter(xs, pred)
  local result = {}
  for _, v in ipairs(xs) do
    if pred(v) then table.insert(result, v) end
  end
  return result
end

return {
  append = append,
  concat = concat,
  length = function(xs) return #xs end,
  reverse = reverse,
  map = map,
  foldl = foldl,
  foldr = function(xs, value, f) return foldl(reverse(xs), value, f) end,
  filter = filter
}
